### PR TITLE
Fix empty stream parsing with ParserStream

### PR DESCRIPTION
**Pull request by stevenvachon**
This needs a test, but don't know which file to put it in.

### CODE REVIEW

**Comment by RReverser** [comment]
Judging from the existing structure, I'd suggest to put one into [test/data/tree_construction_regression](https://github.com/inikulin/parse5/tree/master/test/data/tree_construction_regression) - it will be automatically tested both with regular and streaming mode by [test/fixtures/parser_test.js](https://github.com/inikulin/parse5/blob/e2f5b5872247265431f6a175ab7a290512eac2f5/test/fixtures/parser_test.js#L54-L55).

File: lib/parser/parser_stream.js [hunk 1]
```diff
@@ -28,12 +29,16 @@ inherits(ParserStream, WritableStream);
...
 //WritableStream implementation
 ParserStream.prototype._write = function (chunk, encoding, callback) {
+    this.written = true;
     this.writeCallback = callback;
     this.parser.tokenizer.write(chunk.toString('utf8'), this.lastChunkWritten);
     this._runParsingLoop();
 };
 
 ParserStream.prototype.end = function (chunk, encoding, callback) {
+    if (!this.written && typeof chunk === 'undefined')
```
Comment (RReverser) [inline lib/parser/parser_stream.js]: Btw, `chunk === undefined` would be equivalent and simpler. Or you can even do `chunk = chunk || ''`.

Comment (inikulin) [inline lib/parser/parser_stream.js]: Or even better `chunk === void 0`

Comment (stevenvachon) [inline lib/parser/parser_stream.js]: The lint doesn't allow it.

Comment (inikulin) [inline lib/parser/parser_stream.js]: BTW, I believe it's always safe to write empty string here, so I guess the best way will be to eliminate `this.written` variable and use the second code snippet suggested by Ingvar.

**Comment by stevenvachon** [comment]
The issue solved by this PR should've already been tested by https://github.com/inikulin/parse5/blob/master/test/data/tree_construction/tests1.dat#L839-L846

unless it wasn't ran by `ParserStream`.

**Comment by inikulin** [comment]
@stevenvachon If this issue was covered by this test then why it didn't fail?

**Comment by inikulin** [comment]
This test doesn't test that we return well formed document in that case.

**Comment by stevenvachon** [comment]
@inikulin I'm asking you, since it's your library.

**Comment by stevenvachon** [comment]
It seems to assert the error instead of the document. Is this a bug in the test utils?

**Comment by inikulin** [comment]
We don't even need a test file in this particular case, since it's just an empty string and bug occurs in very specific scenario. Take a look at other regression tests for parser for a clue.

**Comment by stevenvachon** [comment]
If we don't need a test, how are we to ensure that this fix is not undone in the future?

**Comment by inikulin** [comment]
> @inikulin I'm asking you, since it's your library.

idk, it's you claiming that adding new tests is redundant. And I don't understand why, since we don't have tests that cover this particular case. And no, this test shouldn't caught this case, because if you take a look in the code of test utils we'll have a single chunk with `''` content which is written to the parser. Therefore, it doesn't cover our case.

> If we don't need a test, how are we to ensure that this fix is not undone in the future?

I've never claimed that we don't need a test, I wrote that we don't need a test file like this https://github.com/inikulin/parse5/blob/master/test/data/tree_construction/tests1.dat. Since error is caused by the specific combination of API and not by the particular input you'll need to write test in the same format as e.g. this tests: https://github.com/inikulin/parse5/blob/master/test/fixtures/parser_test.js#L62

**Comment by stevenvachon** [comment]
The confusion was "this particular case", where "case" sounded like "this PR". And my questions were asked to save me time, since you're familiar with the code.

This PR should be ready to go now.

**Comment by inikulin** [comment]
@stevenvachon seems like we've had a mutual misunderstanding. LGTM, thank you for this!

**Comment by stevenvachon** [comment]
No problem. Cheers.
